kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,144 @@
+    // Auxiliaries
+    // ###########################################################################
+
+    // Index of first bit that equals one
+    syntax Int ::=  #getFirstOneBit(Int) [function, total]
+
+    rule [gfo-succ]: #getFirstOneBit(X:Int) => log2Int ( X &Int ( ( maxUInt256 xorInt X ) +Int 1 ) ) requires          #rangeUInt(256, X) andBool X =/=Int 0
+    rule [gfo-fail]: #getFirstOneBit(X:Int) => -1                                                    requires notBool (#rangeUInt(256, X) andBool X =/=Int 0)
+
+    // Index of first bit that equals zero
+    syntax Int ::= #getFirstZeroBit(Int) [function, total]
+
+    rule [gfz-succ]: #getFirstZeroBit(X:Int) => #getFirstOneBit ( maxUInt256 xorInt X ) requires         #rangeUInt(256, X)
+    rule [gfz-fail]: #getFirstZeroBit(X:Int) => -1                                      requires notBool #rangeUInt(256, X)
+
+    // Slot updates are performed by the compiler with the help of masks,
+    // which are 256-bit integers of the form 11111111100000000000111111111111111
+    //                                                 |- WIDTH -||-   SHIFT   -|
+
+    // Shift of a mask, in bits and in bytes
+    syntax Int ::= #getMaskShiftBits(Int)  [function, total]
+    syntax Int ::= #getMaskShiftBytes(Int) [function, total]
+
+    rule [gms-bits]: #getMaskShiftBits(X:Int)  => #getFirstZeroBit(X) [concrete(X), simplification]
+
+    rule [gms-bits-succ]: #getMaskShiftBytes(X:Int) => #getFirstZeroBit(X) /Int 8 requires           #getMaskShiftBits(X) modInt 8 ==Int 0   [concrete(X), simplification, preserves-definedness]
+    rule [gms-bits-fail]: #getMaskShiftBytes(X:Int) => -1                         requires notBool ( #getMaskShiftBits(X) modInt 8 ==Int 0 ) [concrete(X), simplification, preserves-definedness]
+
+    // Width of a mask, in bits and in bytes
+    syntax Int ::= #getMaskWidthBits(Int)  [function, total]
+    syntax Int ::= #getMaskWidthBytes(Int) [function, total]
+
+    rule [gmw-bits-succ-1]: #getMaskWidthBits(X:Int) => 256 -Int #getMaskShiftBits(X:Int)             requires  0 <=Int #getMaskShiftBits(X) andBool 0 ==Int X >>Int #getMaskShiftBits(X) [concrete(X), simplification]
+    rule [gmw-bits-succ-2]: #getMaskWidthBits(X:Int) => #getFirstOneBit(X >>Int #getMaskShiftBits(X)) requires  0 <=Int #getMaskShiftBits(X) andBool 0  <Int X >>Int #getMaskShiftBits(X) [concrete(X), simplification]
+    rule [gmw-bits-fail]:   #getMaskWidthBits(X:Int) => -1                                            requires -1 ==Int #getMaskShiftBits(X) [concrete(X), simplification]
+
+    rule [gmw-bytes-succ]: #getMaskWidthBytes(X:Int) => #getMaskWidthBits(X) /Int 8 requires           #getMaskWidthBits(X) modInt 8 ==Int 0   [concrete(X), simplification, preserves-definedness]
+    rule [gmw-bytes-fail]: #getMaskWidthBytes(X:Int) => -1                          requires notBool ( #getMaskWidthBits(X) modInt 8 ==Int 0 ) [concrete(X), simplification, preserves-definedness]
```

Comment:
Could all the `-1` cases be replaced with `[owise]` instead? Maybe that makes it easier to reason about completeness...

Reply:
Or maybe `[owise]` interacts poorly with `concrete`....

Reply:
Ah, hm, so these are simplifications now, but they probably needn't be, since it's again a function definition. If I remove all of the attributes, they could be `[owise]` in the function-rule sense, right?

Reply:
I think so, but I am not sure how `concrete` and `owise` interact, I think it may be poorly. So leaving it as is may be better.

Reply:
Trying with `[owise]` now.

Reply:
Well my concern specifically is that if a rule doesn't apply because of teh `concrete` attribute, perhaps then it will think "then Ishould apply the `owise` case!", and it shouldn't. It should only apply the owise case _if_ it can show that the first case _definitely doesn't_ apply.

Reply:
Yes, so I've removed all `concrete` and `simplification`, now it's just `[owise]` and `[preserves-definedness]`

Reply:
All of these auxiliaries are always called in the lemmas with concrete values, so nothing complex should pop up symbolically.